[쇼핑 장바구니 3, 4단계] 토바에 미션 제출합니다.

tobae-time opened the pull request:
안녕하세요, 제이든!
일단 죄송합니다... 많은 부분을 놓쳐가며 구현을 우선시 했는데요...
상품 목록에서 아이템 목록에 변화가 생길 때마다 여전히 화면 전체가 깜박거립니다.
또한, 최근 본 상품의 경우 상품 상세 페이지에 들어갈 때 보일 때도 있고, 보이지 않을 때도 있습니다. 계속 보였다가 계속 안 보였다가 왔다갔다 하고 있습니다.
okhttp에 대해서는 제대로 이해하고 학습하지 못하여 제가 잘못 구현한 것 같습니다...
상품 목록 화면 앱바의 장바구니 수량에 대해서도 구현하지 못했습니다.
장바구니 화면의 페이지 버튼도 여전히 이상합니다 ㅎㅎ... ㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠ

뷰모델, 라이브데이터, 데이터바인딩를 잘 활용하지 못하는 것 같습니다...

너무너무 부족한 코드지만 잘 부탁드립니다... 죄송합니다 제이든 😭

[이전 코멘트에 대한 추가 코멘트](https://github.com/woowacourse/android-shopping-cart/pull/109#discussion_r2106495457)

# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [ ] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

Commit:
chore: 더미 이미지 리소스 추가

yrsel committed:
feat: 더미 데이터 추가

yrsel committed:
feat: ProductRecyclerView 클릭 리스너 구현

yrsel committed:
feat: ProductListAdapter 구현

yrsel committed:
feat: ProductListViewHolder 구현

yrsel committed:
feat: ProductListActivity 구현

yrsel committed:
feat: Money 일급 객체 구현

Commit:
chore: 불필요한 리소스 제거

Commit:
refactor: main -> product_list로 네이밍 변경

Commit:
feat: 목록에서 아이템 클릭 시 상세 페이지로 이동

Commit:
feat: 장바구니 페이지 xml 구현

yrsel committed:
feat: 장바구니 데이터베이스 구현

yrsel committed:
chore: ktlint 적용

Commit:
feat: 장바구니 담기 버튼을 클릭하면 데이터베이스에 저장하는 기능 추가

yrsel committed:
refactor: BindingAdapter로 금액 표시

yrsel committed:
refactor: CartRepository 패키지 변경

- ProductConverter : 상호 변환 확장 함수
- CartRepository 도메인 계층 이동

yrsel committed:
feat: CartActivity 장바구니 페이지 구현

yrsel committed:
refactor: Appbar -> Toolbar 커스텀

yrsel committed:
fix: CartRepository fetch 안되는 현상 수정

- thread 내부에서 콜백 함수 호출

yrsel committed:
feat: ProductViewModel 뷰모델 생성

Commit:
feat: 상품 상세 페이지 툴바 구현

Commit:
feat: 장바구니 페이지 툴바 구현

Commit:
feat: 삭제 버튼을 누르면 제거하는 기능 추가

Commit:
refactor: ProductViewModel -> CartViewModel 네이밍 변경

Commit:
chore: 더미 데이터 추가

Commit:
refactor: 패키지 구조 변경

Commit:
feat: 상품 목록 더보기 기능 구현

yrsel committed:
refactor: ProductListActivity 더보기 버튼 Span 조절

yrsel committed:
chore: 주석 제거

yrsel committed:
feat: 장바구니 페이지네이션 구현

yrsel committed:
refactor: CartActivity 내부 구조 리팩토링

yrsel committed:
refactor: ProductListActivity 내부 구조 리팩토링

yrsel committed:
chore: ktlint 적용

Commit:
test: UI 테스트 코드 추가

Commit:
refactor: Database 초기화 위치 변경

Commit:
refactor: ProductList -> FashionProductList 네이밍 변경

Commit:
feat: 상품 수량 선택 버튼 추가

Commit:
feat: 상품 수량 선택 버튼 추가

Commit:
refactor: Product가 상품 번호를 갖도록 변경

Commit:
feat: 장바구니 수량 테이블 생성

Commit:
feat: 수량 변경 UI 추가

Commit:
refactor: ProductEntity가 상품 번호를 갖도록 변경

Commit:
feat: 장바구니에서 수량을 선택 연결

Commit:
feat: 수량 변경하는 클릭 리스너 구현

Commit:
fix: 수량이 기존 수량만큼씩 추가로 업데이트 되는 현상 수정

Commit:
fix: 리스너 반대로 동작하는 현상 수정

Commit:
feat: 장바구니에서 수량을 변경하는 기능 추가

Commit:
feat: 상품 상세에서 수량을 변경하는 기능 추가

Commit:
feat: 상품 목록에서 + 버튼을 누르면 장바구니에 반영하는 기능 추가

Commit:
feat: 상품 목록에서 + 버튼을 누르면 수량 변경 뷰가 보이고 장바구니에도 반영되는 기능 추가

Commit:
docs: README.md 4단계 기능 목록 추가

Commit:
feat: 최근 본 상품에 대한 UI, 최근 본 상품을 DB에 저장하는 기능 추가

Commit:
feat: 최근 본 상품이 있을 때, 상품 목록 상단에 아이템이 보이는 기능 추가

Commit:
style: ktlint 적용

Commit:
feat: 최근 본 상품이 있을 때, 상세 페이지의 하단에 마지막에 본 상품을 보여주는 기능 추가

Commit:
feat: 상세 페이지의 하단에서 최근 본 상품을 눌렀을 때, 이전 상세 페이지가 아닌 상품 목록을 보여주는 기능 추가

Commit:
feat: HTTP Client 구현

Commit:
test: MockWebServer를 사용한 테스트 추가

Commit:
refactor: 개수를 요청하도록 변경

Commit:
refactor: 뷰 디테일 수정

Commit:
style: ktlint 적용

Commit:
refactor: 프로젝트 구조 변경

Commit:
refactor: 프로젝트 구조 변경

Commit:
refactor: 로컬 데이터베이스 추상화

KwonDae commented on app/src/main/java/woowacourse/shopping/data/remote/ProductHttpClient.kt:
```diff
@@ -0,0 +1,39 @@
+
+        val body = response.body?.string() ?: throw Exception("응답 없음")
+
+        val jsonArray = JSONArray(body)
+        val products = mutableListOf<Product>()
+
+        for (i in 0 until jsonArray.length()) {
+            val item = jsonArray.getJSONObject(i)
+            val id = item.getLong("id")
+            val imageUrl = item.getString("imageUrl")
+            val name = item.getString("name")
+            val price = item.getInt("price")
+            products.add(Product(id, imageUrl, name, Money(price)))
+        }
```
이부분이 어려우셨군요..!
MockWebServer는 직접 json 응답을 설정해줄 수 있도록 해주는 도구입니다.
지금은 MockWebServer를 호출하지만, 내부적으로는 Product를 리턴해주고 있는것에 불과한데요.

MockServer를 이용하여 상품목록을 불러와야 합니다.
OkHttp 에서 제공하는 MockWebServer 에서
[enqueue](https://square.github.io/okhttp/3.x/mockwebserver/okhttp3/mockwebserver/MockWebServer.html#enqueue-okhttp3.mockwebserver.MockResponse-) 와 [MockResponse](https://square.github.io/okhttp/3.x/mockwebserver/okhttp3/mockwebserver/MockResponse.html) 를 활용할 수 있을거에요.

다음 요청에 반영할 수 있을까요?

KwonDae replied:
추가로 여기에 직렬화/역질렬화 개념도 필요할 수 있어요.
함께 찾아보시면 도움될 것 같습니다!

KwonDae commented on app/src/main/java/woowacourse/shopping/data/repository/CartItemRepositoryImpl.kt:
```diff
@@ -0,0 +1,96 @@
+
+class CartItemRepositoryImpl private constructor(
+    private val dao: CartItemDao,
+    private val cartDao: CartDao,
```
Repository는 데이터의 출처(어디서 데이터가 들어오는지) 까지 관심이 있어야 할까요?
아니면 데이터를 가져오는 것 에만 관심이 있을까요?

현재는 데이터 소스가 Dao 하나이지만, 여러 API, SharedPreference 등 파이프라인이 여러개가 생긴다면 Repository는 비대해질 수 있을거에요.

이러한 구조를 어떻게 설계하면 좋을지 [문서](https://developer.android.com/topic/architecture/data-layer?hl=ko)를 읽어보시고, 개선이 가능하다면 도전해보셔도 좋습니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/ProductListViewModel.kt:
```diff
@@ -0,0 +1,190 @@
+) : ViewModel() {
+    private val cartRepository = CartRepositoryImpl.get()
+    private val cartItemRepository = CartItemRepositoryImpl.get()
+    private val historyRepository = HistoryRepositoryImpl.get()
```
현재 구조에서 ViewModel 테스트가 가능할까요?
지금은 내부 더미데이터를 사용하기에 가능하지만, 외부 API를 사용한다고 하면 테스트가 불가능하며, Repository 구현체와 강하게 결합되어 있는 상태로 변경사항에 영향을 받게 될거에요.

우리가 interface를 사용하는 이유를 생각해보시면서 [문서](https://developer.android.com/codelabs/android-room-with-a-view-kotlin#12)를 참고해보시어 개선해볼 수 있을까요?

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/ProductListViewModel.kt:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    fun loadCartItems() {
+        thread {
```
thread를 ViewModel에서 생성하기 보다, repository에서 생성하여 
콜백으로 응답값을 받도록 구성할 수 있을까요?

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/RecentProductAdapter.kt:
```diff
@@ -0,0 +1,30 @@
+import woowacourse.shopping.databinding.ItemRecentProductBinding
+import woowacourse.shopping.domain.product.Product
+
+class RecentProductAdapter(private val products: List<Product>) : RecyclerView.Adapter<RecentProductViewHolder>() {
```
시간이 된다면 ListAdapter를 적용해보고, 어떤 차이점이 있는지 공부해보셔도 좋습니다

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/RecentProductLayoutViewHolder.kt:
```diff
@@ -0,0 +1,21 @@
+    fun bind(item: List<Product>) {
+        binding.rvRecentProduct.adapter = RecentProductAdapter(item)
+        binding.rvRecentProduct.layoutManager =
+            GridLayoutManager(
```
최근 본 상품은 LinearLayoutManager 를 사용해도 될 것 같아요!

tobae-time replied:
[refactor: 최근 상품 레이아웃 매니저 수정](https://github.com/woowacourse/android-shopping-cart/pull/137/commits/a6764f128eb7cd0b6e947cd7d717f7db736cd949)
 해당 부분을 반영한 커밋입니다! LinearLayoutManager을 사용하여 HORIZONTAL 속성만 주는 게 훨씬 좋은 것 같습니다! 👍

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/FashionProductListAdapter.kt:
```diff
@@ -6,18 +6,23 @@ import android.view.ViewGroup
             is ProductListViewType.LoadMoreType -> R.layout.load_more_item
+            is ProductListViewType.RecentProducts -> R.layout.item_recent_layout
         }
     }
```
ViewType은 토바에가 각 뷰타입마다 설정해주는 상수로 주입해주는게 좋습니다.
layout Resource ID로 고유한 번호이긴 하지만 우리가 예측이 가능한 id가 아닙니다.

우리가 설정할 수 있는 ViewType을 설정해주는게 어떨까요?

KwonDae replied:
또한, 여러 타입의 List를 표현하기 위해 Adapter 내 Adapter가 존재하게 구성해주셨네요!
ViewType 에 따라 중첩된 RecyclerView를 적용하다 보면 depth가 깊어지고, 구조가 복잡해지는 경우가 많습니다.

Recyclerview 1.2.0 버전부터 추가 된 [ConcatAdapter](https://developer.android.com/reference/androidx/recyclerview/widget/ConcatAdapter) 라는 키워드도 존재하는데요
어떤 차이점이 있는지 찾아보셔도 좋습니다 👍

tobae-time replied:
각 어댑터의 ViewType 충돌하지 않도록 해줍니다. 그렇기 때문에 adapter 간 ViewHolder의 재사용이 불가능합니다. 또한 notifyItemInserted(), notifyItemChanged() 호출 시 ConcatAdapter가 자동으로 보정해줍니다.
어댑터 내에 어댑터가 존재할 경우, 불필요한 중첩이 있는 것 같습니다!

저는 상수로 주입해주는 것보다 layout Resource ID로 하였을 때 코드가 직관적인 것 같다고 생각하였습니다. layout ID를 ViewType으로 쓰면 타입과 레이아웃이 명확히 1:1로 대응된다고 생각했는데, 혹시 아닐까요? 내부 ID가 예측 불가능한 숫자라도, 실제로는 R.layout.xxx처럼 이름으로 다루기 때문에 뷰 타입과 연결된 레이아웃을 바로 파악할 수 있어서 좋았습니다. 또한 IDE의 도움을 받아 해당 레이아웃으로 이동할 수 있다는 것도 장점이라고 생각하여 사용하게 되었습니다. 그래서 현재는 layout ID를 직접 사용하는 쪽이 더 직관적으로 느껴지는데, 혹시 이 방식의 단점이나 위험한 케이스가 있을까요?

KwonDae replied:
1. XML의 컴파일 시점의 Int값으로 변환될 때의 값은 우리가 컨트롤 할 수 있는 값이 아닌 예측도 불가합니다. 
2. 현재 명시된 함수는 getViewType 으로 어댑터 내의 뷰 유형으로 해당 XML(R.layout.xxx)은 onCreateViewHolder 에서 사용되는게 역할에 맞다고 생각합니다.

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/fashionlist/FashionProductListActivity.kt:
```diff
@@ -72,10 +72,24 @@ class FashionProductListActivity : AppCompatActivity() {
+
+                        override fun onDecreaseClick(cartItem: CartItem) {
+                            viewModel.decreaseQuantity(cartItem.id)
+                        }
```
비즈니스로직은 viewModel이 담당하고 있는 만큼,
ViewModel에서 리스너를 구현하는건 어떨까요?

KwonDae commented on app/src/main/java/woowacourse/shopping/ui/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,110 @@
+            size = cartRepository.getSize()
+        }
+        update()
+    }
```
전체적으로 ViewModel에서 thread를 생성하여 repository 데이터를 송수신 하고 있습니다.

사실 데이터에 접근하는 모든 로직 관심사는 repository라고 생각이 되는데요
이를 위해 thread를 생성하는 것도 repository에서 해야할 일이라고 추론할 수 있어요.

해당 thread에서 작업이 끝났을 때 리턴값을 전달하는건 그동안 다룬 콜백으로 전달해볼 수 있는데요
이부분은 코루틴을 다루기 전에 알고 넘어가셨으면 합니다 👍

KwonDae reviewed (changes requested):
토바에 안녕하세요!
전체적으로 난이도가 올라가면서 시간이 부족하셨던거 같아요.
다음 미션을 본격적으로 시작하기 전 살펴보고 넘어갔으면 하는 부분들에 대해 코멘트를 남겨드렸으니,
천천히 확인해주시고 요청 부탁드립니다.

Commit:
refactor: 로컬 데이터베이스 추상화

Commit:
refactor: 뷰모델 팩토리 추가 및 수정

Commit:
refactor: 뷰 모델에 상품 목록 로직 통합

Commit:
refactor: ProductEntity 삭제

Commit:
refactor: 수량 변경 XML 파일 수정

Commit:
refactor: 뷰모델 직접 참조로 변경

Commit:
refactor: DetailViewModel DataBinding 적용

- DetailActivity DataBinding ViewModel로 통합
- QuantityClickListener DetailViewModel에서 구현
- ProductDetailActivityTest 주석 처리

Commit:
feat: 수량 변경 시 뷰모델에 위임

Commit:
refactor: 상품 목록 UI 상태 관리 개선

Commit:
fix: 장바구니 화면에서 장바구니 아이템이 보이도록 수정

Commit:
refactor: 장바구니 상태를 onResume 시점에서도업데이트하도록 변경

Commit:
fix: 최근 본 상품이 현재 상품과 중복되는 현상 수정

Commit:
refactor: 최근 상품 레이아웃 매니저 수정

tobae-time reviewed (commented):
안녕하세요, 제이든!
구조 변경과 비동기에 자신이 없어 커밋하지 않고 진행하였습니다... 😭
파일별로 커밋을 나눠보았지만 더욱 보기 힘들어진 것 같아서 해당 부분은 [데이터 레이어를 활용한 구조 변경 및 콜백을 활용하여 응답값을 받도록 변경한 커밋들](https://github.com/woowacourse/android-shopping-cart/pull/137/files/6b0d644dd5abfcafa3ddb6ef760a5706f8a128a8..bc4556474a999b590f5a23957a8dd78003ef3093)을 확인해주시면 감사하겠습니다!
Mock서버는 반영하지 못하였습니다. 해당 부분은 방학 때 보충하여 따로 진행해보겠습니다 ㅜㅜ 
배려해주셔서 감사합니다, 제이든!

KwonDae reviewed (approved):
마지막까지 피드백 반영하시느라 고생많으셨습니다.
미처 반영하지 못한 부분은 현재 진행중인 미션 혹은 이후에 다시 시도해보시기를 바랄게요.
고생 많으셨고 마지막까지 응원하겠습니다 💪
